src/write_influxdb_udp.c

```diff
@@ -0,0 +1,701 @@
+    if ((string[src_pos] == '\\') || (string[src_pos] == ' ') ||
+        (string[src_pos] == ',') || (string[src_pos] == '=') ||
+        (string[src_pos] == '"')) {
+      BUFFER_ADD('\\');
+      BUFFER_ADD(string[src_pos]);
+    } else
+      BUFFER_ADD(string[src_pos]);
+  } /* for */
+  buffer[dst_pos] = 0;
+
+#undef BUFFER_ADD
+
+  return 0;
+} /* }}} int wifxudp_escape_string */
+
+static int write_influxdb_point(char *buffer, int buffer_len,
+                                const data_set_t *ds, const value_list_t *vl) {
+  int status;
+  int offset = 0;
+  gauge_t *rates = NULL;
+  char temp[512] = "";
+  bool have_values = false;
+
+  assert(0 == strcmp(ds->type, vl->type));
+
+  memset(buffer, '\0', buffer_len);
+
+#define BUFFER_ESCAPE(...)                                                     \
+  do {                                                                         \
+    status = wifxudp_escape_string(temp, sizeof(temp), __VA_ARGS__);           \
+    if (status < 0)                                                            \
+      return -1;                                                               \
+  } while (0)
+
+#define BUFFER_ADD(...)                                                        \
+  do {                                                                         \
+    status = snprintf(buffer + offset, buffer_len - offset, __VA_ARGS__);      \
+    if ((status < 1) || (status >= (buffer_len - offset))) {                   \
+      sfree(rates);                                                            \
+      return -1;                                                               \
+    }                                                                          \
+    offset += status;                                                          \
+  } while (0)
+
+  BUFFER_ESCAPE(vl->plugin);
+  BUFFER_ADD("%s", temp);
+
+  BUFFER_ESCAPE(vl->host);
+  BUFFER_ADD(",host=%s", temp);
+  if (strcmp(vl->plugin_instance, "") != 0) {
+    BUFFER_ESCAPE(vl->plugin_instance);
+    BUFFER_ADD(",instance=%s", temp);
+  }
+  if (strcmp(vl->type, "") != 0) {
+    BUFFER_ESCAPE(vl->type);
+    BUFFER_ADD(",type=%s", temp);
+  }
+  if (strcmp(vl->type_instance, "") != 0) {
+    BUFFER_ESCAPE(vl->type_instance);
+    BUFFER_ADD(",type_instance=%s", temp);
+  }
+
+  BUFFER_ADD(" ");
+  for (size_t i = 0; i < ds->ds_num; i++) {
+    if ((ds->ds[i].type != DS_TYPE_COUNTER) &&
+        (ds->ds[i].type != DS_TYPE_GAUGE) &&
+        (ds->ds[i].type != DS_TYPE_DERIVE) &&
+        (ds->ds[i].type != DS_TYPE_ABSOLUTE)) {
+      sfree(rates);
+      return -1;
+    }
+
+    if (ds->ds[i].type == DS_TYPE_GAUGE) {
+      if (isnan(vl->values[i].gauge))
+        continue;
+      if (have_values)
+        BUFFER_ADD(",");
+      BUFFER_ADD("%s=%lf", ds->ds[i].name, vl->values[i].gauge);
+      have_values = true;
+    } else if (wifxudp_config_store_rates) {
+      if (rates == NULL)
```

Comment:
I slightly misunderstand this piece of code. What is the goal?
The `csv` plugin (where all it started) documentation says: `If set to B<true>, convert 'counter' values to rates. If set to B<false> (thedefault) counter values are stored as is`. But `csv` and  all its code copy-pasts  touched all types except GAUGE, not only COUNTER. There is mismatch between implementation and documentation which is misleading.

Reply:
Cite from https://mailman.verplant.org/pipermail/collectd/2014-April/006164.html

```
One attention point is that,
based on CSV plugin, I followed a rule of only calculating rates for
values which are not of GAUGE type. I don't fully understand the
reason for that, but once the data is the same for any output plugin,
I assumed it was right for Network too.
```

1) Can you work on documentation update?

https://mailman.verplant.org/pipermail/collectd/2014-April/006164.html

```
One attention point is that,
based on CSV plugin, I followed a rule of only calculating rates for
values which are not of GAUGE type. I don't fully understand the
reason for that, but once the data is the same for any output plugin,
I assumed it was right for Network too.
```

2) 

Do you have much metrics in your system? 

I have proposals to change `StoreRate` implementations, based on #3159 introduced callback, or to add new `write_rate` callback - that all is to avoid `uc_get_rate` call (which does search for value in cache while we can get value w/o such search).

That would be good to make CPU usage compare, if you have much metrics :-)
That will argue changes.

Reply:
Updated documentation to include derive and absolute. Thank you.

Reply:
>If set to B<true>, convert absolute, counter and derive values to rates.

heh, that is not so such simple change. ))
That is not true - derive is not converted to rate again. Derive is already rate.

Reply:
Please review this my opinion:

```When StoreRate is false, then raw metric values sent via wire.
For DERIVE types, what means raw value is sent, not its change, like for GAUGE, COUNTER and ABSOLUTE.

If StoreRate is true, then change of metric value is sent for DERIVE, COUNTER and ABSOLUTE types - i.e. rate.
GAUGE type always sent as-is.```

You should fully understand how things works (look into `src/daemon/utils_cache.c uc_update()` code where `values_gauge` calculated and do review. )

Reply:
> That is not true - derive is not converted to rate again. Derive is already rate.
> (my this comment is not true too ... ))) )

I've looked at uc_update at utils_cache.c but I'm probably missing something.

`case DS_TYPE_DERIVE: {
      derive_t diff = vl->values[i].derive - ce->values_raw[i].derive;

      ce->values_gauge[i] =
          ((double)diff) / (CDTIME_T_TO_DOUBLE(vl->time - ce->last_time));
      ce->values_raw[i].derive = vl->values[i].derive;
    } break;` 

> So, from my POV, documentation needs more explanations about Collectd internals.

Sure, that's true. I think it's a common problem in several write plugins. I've looked at csv and write_tsdb and the documentation and implementation also differs.

Reply:
>I've looked at uc_update at utils_cache.c but I'm probably missing something.

Let's discuss. Ask your questions )

Reply:
>     2. Another question. Do you have much metrics in your system?

Several thousands with an interval of 10. Metrics with rates are the usual ones: cpu, interface, and some custom made (the last not in all clients).

> That would be good to make CPU usage compare, if you have much metrics :-)
> That will argue changes.

Sure, we can see it.

Reply:
Please take a look:
https://github.com/collectd/collectd/pull/3164/files

That is a draft, just to remove cache scan.
It uses two malloc+memcpy to provide values for write plugin.
If we introduce new callback or option to `plugin_register_write()`, we can avoid one of them.

that is draft, completely unchecked (only compilation succeded), so I can miss something with pointers and sizes, please be patient and fix that if found.

Reply:
> Please review this my opinion:
> 
> ```
> When StoreRate is false, then raw metric values sent via wire.
> For DERIVE types, what means raw value is sent, not its change, like for GAUGE, 
> COUNTER and ABSOLUTE.
> 
> If StoreRate is true, then change of metric value is sent for DERIVE, COUNTER 
> and ABSOLUTE types - i.e. rate. GAUGE type always sent as-is.
> ```

I committed this, but I skipped

>```
>For DERIVE types, what means raw value is sent, not its change, like for GAUGE, 
> COUNTER and ABSOLUTE.
>```

Just because I didn't understand it properly. Maybe we should change the option `StoreRates` for something like:

```
StoreAsGauges. If true store derive, counter and absolute types as gauges.
```

> Please take a look:
> https://github.com/collectd/collectd/pull/3164/files

I did a quick test:
kernel: traps: reader#2[25562] general protection ip:7f14b82aa59c sp:7f14b065a588 error:0 in libc-2.17.so[7f14b8225000+1c2000]
systemd: collectd.service: main process exited, code=killed, status=11/SEGV

not enough time to debug it yet.

Reply:
This needs `gdb` ...
I checked that code on `csv` plugin, and it worked. 
Did you update all binaries and libraries to new versions?

Reply:
>I committed this, but I skipped
That is ok. )
>Just because I didn't understand it properly. 

Setup output to `csv` files and look for stored values.
With 'StoreRate false' you will see 'raw' values. 
The same values are sent "to wire" by write plugins.

>StoreAsGauges. If true store derive, counter and absolute types as gauges.

Maybe, I don't understand word "gauge" properly. Sorry, I'm not native speaker.

If true, then Derive will be sent/stored as derive, really )
If false, Derive is sent as 'counter' which is raw value.

And derive of that value calculated/displayed later. For example, rrd database stores raw values.
That may also be good, if your time-series-database saves these raw values:  while it displays chart based on calculated derive, it also can show raw values for some point of time.

---
>Maybe we should change the option StoreRates for something like: StoreAsGauges.

I don't think such renaming is good idea. It is better to have option name unified between plugins.

Reply:
> This needs `gdb` ...
> I checked that code on `csv` plugin, and it worked.
> Did you update all binaries and libraries to new versions?

Oh, I see... all plugins need to be recompiled and copied.

Here are user and system cpu times for collectd process:

![Screenshot from 2019-05-26 18-50-49](https://user-images.githubusercontent.com/20086694/58384835-cb825500-7fe7-11e9-94c8-f77cde7db52b.png)

values before gap are from master branch and after gap from store-rate branch. The last one seems a bit higher.

Overall cpu usage seems unaffected (server has 2x12 physical cores)

Reply:
>The last one seems a bit higher.

Ouch, unexpectedly...

Reply:
> Turns out that introduced additional 'malloc+memcpy' (which can be avoided, though) takes more cpu than cache search.

Probably a more extensive testing is needed.

Do you think this PR is suitable/ready for commit? 

On the testing side I'm running it successfully ( n * 1000 hosts ==> collectd == loopback, packetsize 16k ==> influxdb) and collectd memory is plain at 800MB.